ext/php7/ddtrace.c

```diff
@@ -47,34 +50,58 @@
 
 bool ddtrace_has_excluded_module;
 
-atomic_int ddtrace_first_rinit;
 atomic_int ddtrace_warn_legacy_api;
 
 ZEND_DECLARE_MODULE_GLOBALS(ddtrace)
 
 PHP_INI_BEGIN()
 STD_PHP_INI_BOOLEAN("ddtrace.disable", "0", PHP_INI_SYSTEM, OnUpdateBool, disable, zend_ddtrace_globals,
                     ddtrace_globals)
-#if _BUILD_FROM_PECL_
-STD_PHP_INI_ENTRY("ddtrace.request_init_hook", "@php_dir@/datadog_trace/bridge/dd_wrap_autoloader.php", PHP_INI_SYSTEM,
-                  OnUpdateString, request_init_hook, zend_ddtrace_globals, ddtrace_globals)
-#else
-STD_PHP_INI_ENTRY("ddtrace.request_init_hook", "", PHP_INI_SYSTEM, OnUpdateString, request_init_hook,
-                  zend_ddtrace_globals, ddtrace_globals)
-#endif
 
 // Exposed for testing only
 STD_PHP_INI_ENTRY("ddtrace.cgroup_file", "/proc/self/cgroup", PHP_INI_SYSTEM, OnUpdateString, cgroup_file,
                   zend_ddtrace_globals, ddtrace_globals)
 PHP_INI_END()
 
+#if PHP_VERSION_ID >= 70300 && PHP_VERSION_ID < 70400
+static void ddtrace_sort_modules(void *base, size_t count, size_t siz, compare_func_t compare, swap_func_t swp) {
+    UNUSED(siz);
+    UNUSED(compare);
+    UNUSED(swp);
+
+    for (Bucket *module = base, *end = module + count, *ddtrace_module = NULL; module < end; ++module) {
+        zend_module_entry *m = (zend_module_entry *)Z_PTR(module->val);
+        if (m->name == ddtrace_module_entry.name) {
+            ddtrace_module = module;
+        }
+        if (ddtrace_module && strcmp(m->name, "Zend OPcache") == 0) {
+            Bucket tmp = *ddtrace_module;
+            *ddtrace_module = *module;
+            *module = tmp;
+            break;
+        }
+    }
+}
+#endif
+
 static int ddtrace_startup(struct _zend_extension *extension) {
+#if PHP_VERSION_ID >= 70300 && PHP_VERSION_ID < 70400
+    // Turns out with zai config we have dynamically allocated INI entries. This does not play well with PHP 7.3
+    // As of PHP 7.3 opcache stores INI entry values in SHM. However, only as of PHP 7.4 opcache delays detaching SHM.
+    // In PHP 7.3 SHM is freed in MSHUTDOWN, which may be executed before our extension, if we do not force an order.
+    // We have to sort this manually here, as opcache only registers itself as extension during zend_extension.startup.
```

Comment:
> [...] opcache only registers itself as extension during zend_extension.startup.

You are referring here to the fact that opcache is a hybrid module + extension, or something else?

Reply:
correct